Core/include/Acts/TrackFitting/Chi2Fitter.hpp

```diff
@@ -0,0 +1,757 @@
+          chi2FitterOptions.geoContext, chi2FitterOptions.magFieldContext,
+          logger);
+
+      // Set the trivial propagator options
+      propOptions.setPlainOptions(chi2FitterOptions.propagatorPlainOptions);
+
+      // Catch the actor and set the measurements
+      auto& chi2Actor = propOptions.actionList.template get<Chi2Actor>();
+      chi2Actor.inputMeasurements = &inputMeasurements;
+      chi2Actor.multipleScattering = chi2FitterOptions.multipleScattering;
+      chi2Actor.energyLoss = chi2FitterOptions.energyLoss;
+      chi2Actor.freeToBoundCorrection = chi2FitterOptions.freeToBoundCorrection;
+      chi2Actor.extensions = chi2FitterOptions.extensions;
+
+      typename propagator_t::template action_list_t_result_t<
+          CurvilinearTrackParameters, Actors>
+          inputResult;
+
+      auto& r = inputResult.template get<Chi2FitterResult<traj_t>>();
+
+      if (trajectory) {
+        r.fittedStates = trajectory;
+      } else {
+        r.fittedStates = std::make_shared<traj_t>();
+      }
+
+      using paramType = typename std::conditional<
+          std::is_same<start_parameters_t, parameters_t>::value,
+          std::variant<start_parameters_t>,
+          std::variant<start_parameters_t, parameters_t>>::type;
+      paramType vParams = sParameters;
+      // start_parameters_t and parameter_t can be the same
+
+      auto result = m_propagator.template propagate(sParameters, propOptions,
+                                                    std::move(inputResult));
+
+      if (!result.ok()) {
+        ACTS_ERROR("chi2 | it=" << i
+                                << " | propapation failed: " << result.error());
+        return result.error();
+      }
+
+      Chi2Result c2rCurrent =
+          std::move(result.value().template get<Chi2Result>());
+
+      /// It could happen that the fit ends in zero measurement states.
+      /// The result gets meaningless so such case is regarded as fit failure.
+      if (c2rCurrent.result.ok() and not c2rCurrent.measurementStates) {
+        c2rCurrent.result = Result<void>(Chi2FitterError::NoMeasurementFound);
+      }
+
+      if (!c2rCurrent.result.ok()) {
+        ACTS_ERROR("chi2 | it=" << i << " | Chi2Fitter failed: "
+                                << c2rCurrent.result.error() << ", "
+                                << c2rCurrent.result.error().message());
+        return c2rCurrent.result.error();
+      }
+
+      c2rCurrent.residuals =
+          Eigen::Map<ActsDynamicVector>(c2rCurrent.collectorResiduals.data(),
+                                        c2rCurrent.collectorResiduals.size());
+      // TODO: is this safe? from Stackoverflow: "dangerous! Because the Eigen
+      // object will NOT create its own memory. It will operate on the memory
+      // provided by "data". In other words, working with the Eigen object
+      // when the "data" object is out of scope will result in a segmentation
+      // fault (or memory access violation)."
```

Comment:
should be safe as long as you can guarantee for the life time of the other object